zowe/zowe-cli
# fix(secrets): Reduce keychain unlock prompts on MacOS

**What It Does**

Reduces the number of keychain unlock prompts on MacOS for simultaneous access to secrets by multiple instances of the same application.

As a security feature, MacOS always re-prompts for keychain access after a secret has been modified by another process. This is why after running `zowe config secure` which updates credentials using `node`, you get a prompt for Zowe Explorer (the VS Code extension host process) to access credentials.

Suppose there are 3 VS Code instances open: A, B, and C. Currently when a vault change event is triggered, all 3 instances attempt to access the keychain simultaneously and Mac users will see 3 keychain prompts (ouch).

By adding a mutex associated with the calling process, we can make B & C wait to access credentials until A has finished accessing them. Now Mac users will see just 1 keychain prompt (0 would be ideal but 1 is way better than 3).

**How to Test**

This PR can only be tested on MacOS since the issue is OS-specific:
1. Download the `keyring.node` artifact and copy it into your ZE prebuilds folder.
2. Dequarantine the file with the command `xattr -d com.apple.quarantine *.node`
3. Close all VS Code windows and re-open multiple of them.
4. Run the `zowe config secure` command and press Enter repeatedly to skip all prompts.
5. Notice that only 1 prompt is shown for keychain access.

**Review Checklist**
I certify that I have:


**Additional Comments**
